keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Volume Scale Testing Plan
+authors:
+  - "@msau42"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-scalability
+reviewers:
+  - "@pohly"
+  - "@saad-ali"
+  - "@wojtek-t"
+approvers:
+  - "@saad-ali"
+  - "@wojtek-t"
+editor: TBD
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Volume Scale Testing Plan
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [WIP Test Environment](#wip-test-environment)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability of K8s storage components.
+
+## Motivation
+
+Adding storage scale tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide a framework that vendors can use to run scale tests against their CSI drivers.
```

Comment:
Can you explain that in a bit more details? I would like to see the expected result of it (i.e. what a endor would have to do to test that).
So let's say that I would like to test exactly the same scenarios that we will create tests for, but instead of say using mock driver, use e.g. GCE PD. I think that the ideal situation is to just:
- configure a cluster correctly (i guess, run it on GCE should be enough here)
- use appropriate PV type in tests
And that's all.
Or do you have something else on your mind?

Reply:
Before running the tests, a vendor would need to:
* define a storageclass
* launch their csi driver

The main requirement I wanted to convey here is that the tests should be written so that we don't need to hard code the volume type.  Maybe we can just require that the default storageclass is set appropriately before the test.  I will try to reword this a bit.